src/ast/ast_pp_dot.cpp

```diff
@@ -0,0 +1,132 @@
+/*++
+
+Abstract: Pretty-printer for proofs in Graphviz format
+
+--*/
+
+#include "util/util.h"
+#include "util/map.h"
+#include "ast_pp_dot.h"
+
+// string escaping for DOT
+std::string escape_dot(std::string const & s) {
+    std::string res;
+    for (auto c : s) {
+        if (c == '\n')
+            res.append("\\l");
+        else
+            res.push_back(c);
+    }
+    return res;
+}
+
+// map from proofs to unique IDs
+typedef map<const expr*, unsigned, obj_ptr_hash<const expr>, default_eq<const expr*> > expr2id;
+typedef map<const expr*, bool, obj_ptr_hash<const expr>, default_eq<const expr*> > set_expr;
+
+// temporary structure for traversing the proof and printing it
+struct ast_pp_dot_st {
+    const ast_pp_dot *      m_pp;
+    set_expr                m_printed;
+    expr2id                 m_id_map;
+    svector<const expr *>   m_to_print;
+    std::ostream &          m_out;
+    unsigned                m_next_id;
+    bool                    m_first;
+
+    ast_pp_dot_st(const ast_pp_dot * pp, std::ostream & out) :
+        m_pp(pp),
+        m_out(out),
+        m_next_id(0),
+        m_id_map(),
+        m_to_print(),
+        m_printed(),
+        m_first(true) {}
+    
+    void push_term(const expr * a) { m_to_print.push_back(a); }        
+
+    void pp_loop() {
+        // DFS traversal
+        auto& m = get_manager();
+        while (!m_to_print.empty()) {
+            const expr * a = m_to_print.back();
+            m_to_print.pop_back();
+            if (!m_printed.contains(a)) {
+                m_printed.insert(a, true);
+                if (m.is_proof(a))
+                    pp_step(to_app(a));
+                else
+                    pp_atomic_step(a);
+            }
+        }
+    }
+
+private:
+
+    ast_manager & get_manager() const { return m_pp->get_manager(); }
+
+    // label for an expression
+    std::string label_of_expr(const expr * e) const {
+        expr_ref er((expr*)e, get_manager());
+        std::ostringstream out;
+        out << er << std::flush;
+        return escape_dot(out.str());
+    }
+
+    void pp_atomic_step(const expr * e) {
+        unsigned id = get_id(e);
+        m_out << "node_" << id << " [shape=box,color=\"yellow\",style=\"filled\",label=\"" << label_of_expr(e) << "\"] ;" << std::endl;
+    }
+    
+    void pp_step(const proof * p) {
+        auto m = get_manager();
```

Comment:
Not sure if this gets a reference or copy.
Suggestion: add an attribute 

               ast_manager& m; 

to the class and use this instead of retrieving "get_manager()".

In the long run, using get_manager() turned out to be a bit verbose.